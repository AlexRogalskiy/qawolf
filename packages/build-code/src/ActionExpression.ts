export class ActionExpression {
  code() {
    return "";
  }

  updatableCode() {
    return this.code();
  }
}
