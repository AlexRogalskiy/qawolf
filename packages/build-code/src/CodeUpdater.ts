import { buildWorkflow } from "@qawolf/build-workflow";
import { logger } from "@qawolf/logger";
import { ElementEvent, Step, Workflow } from "@qawolf/types";
import { sortBy } from "lodash";
import { buildStepsCode } from "./buildStepsCode";
import { getIndentation, indent, removeLineIncludes } from "./indent";

export const CREATE_CODE_SYMBOL = "// 🐺 CREATE CODE HERE";

type ConstructorOptions = {
  device?: string;
  isTest?: boolean;
  name: string;
  url: string;
};

type PrepareStepsOptions = {
  newEvents?: ElementEvent[];
  onlyFinalSteps?: boolean;
};

export class CodeUpdater {
  private _events: ElementEvent[] = [];
  private _options: ConstructorOptions;

  private _pendingStepIndex: number = 0;
  public _steps: Step[] = [];

  constructor(options: ConstructorOptions) {
    this._options = options;
  }

  public static hasCreateSymbol(code: string) {
    return code.includes(CREATE_CODE_SYMBOL);
  }

  private _buildWorkflow(onlyFinalSteps?: boolean) {
    const workflow = buildWorkflow({
      device: this._options.device,
      events: sortBy(this._events, e => e.time),
      onlyFinalSteps,
      name: this._options.name,
      url: this._options.url
    });

    return workflow;
  }

  public getEvents() {
    return this._events;
  }

  public getNumPendingSteps() {
    return this._steps.length - this._pendingStepIndex;
  }

  public prepareSteps(options: PrepareStepsOptions): Workflow {
    if (options.newEvents) {
      this._events.push(...options.newEvents);
    }

    const workflow = this._buildWorkflow(options.onlyFinalSteps);

    const newSteps = workflow.steps.slice(this._steps.length);
    newSteps.forEach(step => {
      logger.debug(`CodeUpdater: new step ${step.action}`);
      this._steps.push(step);
    });

    return workflow;
  }

  public updateCode(code: string, finalize: boolean = false): string {
    if (!CodeUpdater.hasCreateSymbol(code)) {
      throw new Error("Cannot update code without create symbol");
    }

    if (!finalize && this.getNumPendingSteps() < 1) {
      throw new Error("No code to update");
    }

    const startIndex = this._pendingStepIndex;

    // move the pending step forward
    this._pendingStepIndex = this._steps.length;

    let codeToInsert =
      buildStepsCode({
        startIndex,
        steps: this._steps,
        isTest: this._options.isTest
      }) + CREATE_CODE_SYMBOL;

    const numSpaces = getIndentation(code, CREATE_CODE_SYMBOL);

    let updatedCode = code.replace(
      CREATE_CODE_SYMBOL,
      indent(codeToInsert, numSpaces, 1)
    );

    if (finalize) {
      updatedCode = removeLineIncludes(code, CREATE_CODE_SYMBOL);
    }

    return updatedCode;
  }
}
