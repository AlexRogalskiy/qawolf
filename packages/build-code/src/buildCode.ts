import { Workflow } from "@qawolf/types";
import { readFileSync } from "fs-extra";
import { compile } from "handlebars";
import { camelCase } from "lodash";
import { resolve } from "path";
import { formatLaunch } from "./formatLaunch";
import { formatStep } from "./formatStep";

type BuildCodeOptions = {
  test?: boolean;
  workflow: Workflow;
};

const scriptTemplate = compile(
  readFileSync(resolve(__dirname, "../static/script.hbs"), "utf8")
);

const testTemplate = compile(
  readFileSync(resolve(__dirname, "../static/test.hbs"), "utf8")
);

export const buildCode = (options: BuildCodeOptions) => {
  const { workflow } = options;

  const template = options.test ? testTemplate : scriptTemplate;

  const code = template({
    launch: formatLaunch(workflow.url, workflow.device),
    name: camelCase(workflow.name),
    steps: workflow.steps.map((step, i) => {
      const previousStep = i > 0 ? workflow.steps[i - 1] : null;

      return formatStep(step, previousStep);
    }),
    url: workflow.url
  });

  return code;
};
