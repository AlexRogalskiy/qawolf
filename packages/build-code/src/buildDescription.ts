import { Step } from "@qawolf/types";
import { describeDoc } from "@qawolf/web";
import { isNil } from "lodash";

export const buildDescription = (step: Step): string => {
  if (step.action === "scroll") {
    return `scroll`;
  }

  const description = describeDoc(step.html);

  const target = step.html.node;

  // link/input
  const tagName = `${target.name === "a" ? " link" : ` ${target.name}` || ""}`;

  if (step.action === "click") {
    return `click${description}${tagName}`;
  }

  if (step.action === "select") {
    return `select${description}`;
  }

  if (step.action === "type") {
    if (step.replace && (step.value === "" || isNil(step.value))) {
      return `clear${description}${tagName}`;
    }

    const value = step.value as string;
    if (value && value.indexOf("↓Enter") === 0) {
      return `Enter`;
    }

    if (value && value.indexOf("↓Tab") === 0) {
      return `Tab`;
    }

    return `type into${description}${tagName}`;
  }

  throw new Error(`Invalid step action ${step.action}`);
};
