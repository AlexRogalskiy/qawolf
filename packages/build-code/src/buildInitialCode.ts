import { camelCase } from "lodash";
import { formatLaunch } from "./formatLaunch";

export type InitialCodeOptions = {
  createCodeSymbol: string;
  device?: string;
  isTest?: boolean;
  name: string;
  url: string;
};

type TemplateOptions = {
  createCodeSymbol: string;
  name: string;
  launch: string;
};

const buildInitialScript = ({
  name,
  launch,
  createCodeSymbol
}: TemplateOptions) => {
  const code = `const { launch } = require("qawolf");
const selectors = require("../selectors/${name}");

const ${name} = async () => {
  const browser = ${launch}
  ${createCodeSymbol}
  await browser.close();
};

${name}();`;

  return code;
};

const buildInitialTest = ({
  name,
  launch,
  createCodeSymbol
}: TemplateOptions) => {
  const code = `const { launch } = require("qawolf");
const selectors = require("../selectors/${name}");

describe('${name}', () => {
  let browser;

  beforeAll(async () => {
    browser = ${launch}
  });

  afterAll(() => browser.close());
  ${createCodeSymbol}
});`;

  return code;
};

export const buildInitialCode = (options: InitialCodeOptions) => {
  const templateOptions = {
    createCodeSymbol: options.createCodeSymbol,
    launch: formatLaunch(options.url, options.device),
    name: camelCase(options.name)
  };

  if (options.isTest) {
    return buildInitialTest(templateOptions);
  }

  return buildInitialScript(templateOptions);
};
