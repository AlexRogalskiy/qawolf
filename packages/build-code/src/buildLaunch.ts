export const buildLaunch = (url: string, device?: string): string => {
  if (device) {
    return `await launch({ device: "${device}", url: "${url}" });`;
  }

  return `await launch({ url: "${url}" });`;
};
