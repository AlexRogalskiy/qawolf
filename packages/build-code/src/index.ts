export * from "./buildInitialCode";
export { stepToSelector } from "./buildSelector";
export * from "./buildVirtualCode";
export * from "./CodeReconciler";
