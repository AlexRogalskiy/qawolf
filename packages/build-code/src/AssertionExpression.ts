import { ActionExpression } from "./ActionExpression";

type ConstructorOptions = {
  action: ActionExpression;
  description: string;
};

export class AssertionExpression {
  private _action: ActionExpression;
  private _description: string;

  constructor({ action, description }: ConstructorOptions) {
    this._action = action;
    this._description = description;
  }

  private _formatIt() {
    const description = this._description;
    if (!description.includes(`"`)) return `"can ${description}"`;
    if (!description.includes("'")) return `'can ${description}'`;
    return JSON.stringify("can " + description);
  }

  code() {
    const code = `\nit(${this._formatIt()}, async () => {\n  ${this._action.code()}\n  });\n`;
    return code;
  }

  updatableCode() {
    return this._action.updatableCode();
  }
}
