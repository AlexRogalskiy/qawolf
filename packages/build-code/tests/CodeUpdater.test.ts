import { loadEvents } from "@qawolf/fixtures";
import { ElementEvent } from "@qawolf/types";
import { CodeUpdater, CREATE_CODE_SYMBOL } from "../src/CodeUpdater";
import { last, pick } from "lodash";

let events: ElementEvent[];

beforeAll(async () => {
  events = await loadEvents("scroll_login");
});

describe("CodeUpdater.hasCreateSymbol", () => {
  it("returns true when the create symbol is found", () => {
    expect(
      CodeUpdater.hasCreateSymbol(`someCode();\n${CREATE_CODE_SYMBOL}`)
    ).toBe(true);
  });

  it("returns false when the create symbol is missing", () => {
    expect(CodeUpdater.hasCreateSymbol("")).toBe(false);
  });
});

describe("CodeUpdater.prepareSteps", () => {
  it("appends new steps", () => {
    const codeUpdater = new CodeUpdater({ name: "myTest", url: "localhost" });

    let numSteps = 0;

    let stepEvents: any[] = [];

    events.forEach(event => {
      codeUpdater.prepareSteps({ newEvents: [event], onlyFinalSteps: true });

      // track each event that causes a step to be created
      if (codeUpdater._steps.length !== numSteps) {
        stepEvents.push({
          event: pick(event, "name", "value"),
          step: last(codeUpdater._steps)!.action
        });

        numSteps = codeUpdater._steps.length;
      }
    });

    // test the events and new steps align properly
    expect(stepEvents).toMatchSnapshot();
  });
});

describe("CodeUpdater.updateCode", () => {
  // TODO test multiple updates w/ snapshot

  it("replaces the create symbol with new steps", () => {
    for (let isTest of [true, false]) {
      const codeUpdater = new CodeUpdater({
        isTest,
        name: "myTest",
        url: "localhost"
      });

      expect(codeUpdater.getNumPendingSteps()).toEqual(0);
      codeUpdater.prepareSteps({ newEvents: events, onlyFinalSteps: true });
      expect(codeUpdater.getNumPendingSteps()).toEqual(8);

      const codeToUpdate = `myOtherCode();\n${CREATE_CODE_SYMBOL}`;
      const updatedCode = codeUpdater.updateCode(codeToUpdate);

      expect(updatedCode).toMatchSnapshot(
        isTest ? "createTestSteps" : "createScriptSteps"
      );

      expect(codeUpdater.getNumPendingSteps()).toEqual(0);
    }
  });

  it("removes the create symbol line on finalize", () => {
    // TODO..
  });
});
