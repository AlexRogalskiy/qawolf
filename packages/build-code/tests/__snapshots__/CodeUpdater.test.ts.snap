// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeUpdater.prepareSteps appends new steps 1`] = `
Array [
  Object {
    "event": Object {
      "name": "click",
    },
    "step": "click",
  },
  Object {
    "event": Object {
      "name": "click",
    },
    "step": "click",
  },
  Object {
    "event": Object {
      "name": "keydown",
      "value": "Tab",
    },
    "step": "type",
  },
  Object {
    "event": Object {
      "name": "paste",
      "value": "SuperSecretPassword!",
    },
    "step": "type",
  },
  Object {
    "event": Object {
      "name": "keydown",
      "value": "Enter",
    },
    "step": "type",
  },
  Object {
    "event": Object {
      "name": "click",
    },
    "step": "click",
  },
]
`;

exports[`CodeUpdater.updateCode replaces the create symbol with new steps: createScriptSteps 1`] = `
"myOtherCode();
// scroll
await browser.scroll(selectors[0], { x: 0, y: 378 });
// click \\"Form Authentication\\" link
await browser.click(selectors[1]);
// click \\"Username\\" input
await browser.click(selectors[2]);
// type into \\"Username\\" input
await browser.type(selectors[3], \\"tomsmith\\");
// Tab
await browser.type(selectors[4], \\"↓Tab↑Tab\\");
// type into \\"Password\\" input
await browser.type(selectors[5], \\"SuperSecretPassword!\\");
// Enter
await browser.type(selectors[6], \\"↓Enter↑Enter\\");
// click \\"Logout\\" link
await browser.click(selectors[7]);
// 🐺 CREATE CODE HERE"
`;

exports[`CodeUpdater.updateCode replaces the create symbol with new steps: createTestSteps 1`] = `
"myOtherCode();

it('can scroll', async () => {
  await browser.scroll(selectors[0], { x: 0, y: 378 });
});

it('can click \\"Form Authentication\\" link', async () => {
  await browser.click(selectors[1]);
});

it('can click \\"Username\\" input', async () => {
  await browser.click(selectors[2]);
});

it('can type into \\"Username\\" input', async () => {
  await browser.type(selectors[3], \\"tomsmith\\");
});

it('can Tab', async () => {
  await browser.type(selectors[4], \\"↓Tab↑Tab\\");
});

it('can type into \\"Password\\" input', async () => {
  await browser.type(selectors[5], \\"SuperSecretPassword!\\");
});

it('can Enter', async () => {
  await browser.type(selectors[6], \\"↓Enter↑Enter\\");
});

it('can click \\"Logout\\" link', async () => {
  await browser.click(selectors[7]);
});
// 🐺 CREATE CODE HERE"
`;
